### PR TITLE
Add passing record types directly to parameterizedQuery support

### PR DESCRIPTION
**Pull request by lahirusamith**
## Purpose
$title

Fixes:

## Checklist
- [ ] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/DefaultStatementParameterProcessor.java [hunk 1]
```diff
@@ -612,6 +612,13 @@ protected void setXml(Connection connection, PreparedStatement preparedStatement
         preparedStatement.setObject(index, value.getTextValue(), Types.SQLXML);
     }
 
+    @Override
+    protected int setCustomBMapParams(Connection connection, PreparedStatement preparedStatement, int index,
+                                      Object value, boolean returnType) throws DataError, SQLException {
+        throw new DataError("Unsupported type passed in column index: " + index);
+    }
+
+
```
Comment (daneshk) [inline native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/DefaultStatementParameterProcessor.java:621]:
```suggestion
```

**Comment by daneshk** [comment]
@niveathika Can you review this. we need this fix to support [1]

1. ballerina-platform/ballerina-standard-library#1763#issuecomment-908003693

File: native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java [hunk 2]
```diff
@@ -486,4 +491,8 @@ private int getSQLType(BObject typedValue) throws DataError, SQLException {
         return sqlTypeValue;
     }
 
+    private int setBMapParams(Connection connection, PreparedStatement preparedStatement, int index,
+                                               Object value, boolean returnType) throws DataError, SQLException {
+        return setCustomBMapParams(connection, preparedStatement, index, value, returnType);
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java:496]:
```suggestion
        return setCustomBOpenRexord(connection, preparedStatement, index, value, returnType);
```

Comment (lahirusamith) [reply native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java]: corrected in https://github.com/ballerina-platform/module-ballerina-sql/pull/358/commits/cd74b3fe593a9cc709b97184cdbe4d6f82d9aea0

**Review by daneshk — DISMISSED** [review]
LGTM

File: ballerina/tests/simple-params-query-test.bal [hunk 3]
```diff
@@ -581,6 +581,26 @@ function queryTimeStringInvalidParam() {
...
+
+@test:Config {
+    groups: ["query", "query-simple-params"]
+}
+function queryInvalidRecordParam() {
+    InvalidRecord recordValue = {id : 1};
+    ParameterizedQuery sqlQuery = `SELECT * from DataTable WHERE row_id = ${recordValue}`;
+    record {}|error? result = trap queryMockClient(simpleParamsDb, sqlQuery);
+    test:assertTrue(result is error);
+    io:println(result);
```
Comment (kaneeldias) [inline ballerina/tests/simple-params-query-test.bal:596]: Is the print statement necessary?

Comment (lahirusamith) [reply ballerina/tests/simple-params-query-test.bal]: corrected in https://github.com/ballerina-platform/module-ballerina-sql/pull/358/commits/2e7b8240f65198e3eb8a641457944d7b0bdfe7e9

**Review by kaneeldias — APPROVED** [review]

File: native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java [hunk 4]
```diff
@@ -211,6 +214,8 @@ public int setSQLValueParam(Connection connection, PreparedStatement preparedSta
         } else if (object instanceof BXml) {
             setXml(connection, preparedStatement, index, (BXml) object);
             return Types.SQLXML;
+        } else if (object instanceof BMap) {
+            return setBMapParams(connection, preparedStatement, index, object, returnType);
```
[the remark this answers is not in the document]
Comment (lahirusamith) [reply native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java:218]: This s`etBMapParams` method will be used to add future record types support (time:Civil etc). and `setCustomBOpenRecord` is used to provide additional record types support for  extended db connector modules.

[the remark this answers is not in the document]
Comment (niveathika) [reply native/src/main/java/io/ballerina/stdlib/sql/parameterprocessor/AbstractStatementParameterProcessor.java:218]: Ack

**Review by niveathika — APPROVED** [review]

**Pull request merged**